### PR TITLE
Slit trace algorithm

### PR DESCRIPTION
**Pull request by rcooke-ast**
Modified slit trace algorithm to be used for MOS data (DEIMOS+LRIS)
Improved algorithms for slit assignment and edge matching.

### CODE REVIEW

**Comment by adwasser** [comment]
I forgot to ask earlier, but is it worth merging this right now, when there are imminent changes?

**Comment by rcooke-ast** [comment]
I'm hoping to update this PR within a few days. I'll also wait for X to do one pass through. I hope that's ok! You can always branch off this branch (called, say, newbranch), and then later merge the future master branch into newbranch (if that makes sense)?

File: doc/tracing.rst [hunk 1]
```diff
@@ -7,10 +7,53 @@ Tracing
 This document will describe how the code traces the
 edges of slits.
 
+The edge tracing algorithm uses a combination of a
+median and sobel filter to identify significant
```
Comment (profxj) [inline doc/tracing.rst]: filters

File: pypit/armlsd.py [hunk 2]
```diff
@@ -127,7 +127,17 @@ def ARMLSD(argflag, spect, fitsdict, reuseMaster=False, reloadMaster=True):
             if ('trace'+slf._argflag['masters']['setup'] not in slf._argflag['masters']['loaded']):
                 ###############
                 # Determine the edges of the spectrum (spatial)
-                lordloc, rordloc, extord = artrace.trace_orders(slf, slf._mstrace[det-1], det, singleSlit=True, pcadesc="PCA trace of the slit edges")
+                # filename = "/Users/rcooke/Desktop/edge_detect/"
```
Comment (profxj) [inline pypit/armlsd.py]: May be able to remove these lines

File: pypit/artrace.py [hunk 3]
```diff
@@ -315,17 +541,32 @@ def trace_orders(slf, mstrace, det, pcadesc="", maskBadRows=False, singleSlit=Fa
...
+    if msgs._debug['trace']:
+        binbpx = np.zeros(mstrace.shape, dtype=np.int)
+        xs = np.arange(mstrace.shape[slf._dispaxis]*1.0)*slf._spect['det'][det-1]['xgap']
+        xt = 0.5 + np.arange(mstrace.shape[slf._dispaxis]*1.0) + xs
+        ys = np.arange(mstrace.shape[1-slf._dispaxis])*slf._spect['det'][det-1]['ygap']*slf._spect['det'][det-1]['ysize']
+        yt = slf._spect['det'][det-1]['ysize']*(0.5 + np.arange(mstrace.shape[1-slf._dispaxis]*1.0)) + ys
+        xloc, yloc = np.meshgrid(xt, yt)
+        plxbin, plybin = xloc.T, yloc.T
+
+#    msgs.work("binby=1 makes this slow and ineffective -- increase this to 10, and add as a parameter of choice by the user")
```
Comment (profxj) [inline pypit/artrace.py]: Can these be removed?

File: pypit/settings/settings.armlsd [hunk 4]
```diff
@@ -76,14 +76,18 @@ trace comb sat_pix      reject        # What to do with saturated pixels (option
...
 trace orders diffpolyorder  2         # What is the order of the 2D function that should be used to fit the 2d solution for the spatial size of all orders?
 trace orders sigdetect  5.0           # Sigma detection threshold for edge detection
-trace orders fracignore 0.6           # If an order spans less than this fraction over the detector, it will be reconstructed and not fitted
-trace orders pca [3,2,1,0,0,0]        # What order polynomials should be used to fit the principle components
-trace orders pcxpos     3             # How many extra orders to predict in the positive direction
-trace orders pcxneg     3             # How many extra orders to predict in the negative direction
+trace orders fracignore 0.01           # If an order spans less than this fraction over the detector, it will be reconstructed and not fitted
+trace orders pcatype pixel            # Should the PCA be performed using pixel position (pixel) or by spectral order (order). The latter is used for echelle spectroscopy.
+trace orders pcaparams [3,2,1,0]  # What order polynomials should be used to fit the principle components
```
Comment (profxj) [inline pypit/settings/settings.armlsd]: 3 is pretty large.  Is it stable?

**Comment by profxj** [comment]
Just a few inline comments.  I'll admit I'm overwhelmed with code! ;)

Please confirm this branch passes on the test suite and then merge.

**Comment by adwasser** [comment]
Seems to work on Kast, but LRIS tests are failing.
~~~
Running pypit on REDUX_OUT/LRIS_blue/long_400_3400_d560/lris_blue_long_400_3400_d560.pypit ---
Done running pypit on REDUX_OUT/LRIS_blue/long_400_3400_d560/lris_blue_long_400_3400_d560.pypit ---
    status = armlsd.ARMLSD(argflag, spect, fitsdict)
  File "/Users/asher/work/software/PYPIT/pypit/armlsd.py", line 130, in ARMLSD
    lordloc, rordloc, extord = artrace.trace_slits(slf, slf._mstrace[det-1], det, pcadesc="PCA trace of the slit edges")
  File "/Users/asher/work/software/PYPIT/pypit/artrace.py", line 799, in trace_slits
    lmin, lmax = -np.max(edgearr[ww]), -np.min(edgearr[ww])  # min/max are switched because of the negative signs
  File "/usr/local/lib/python2.7/site-packages/numpy/core/fromnumeric.py", line 2268, in amax
    out=out, keepdims=keepdims)
  File "/usr/local/lib/python2.7/site-packages/numpy/core/_methods.py", line 26, in _amax
    return umr_maximum(a, axis, None, out, keepdims)
ValueError: zero-size array to reduction operation maximum which has no identity
Running pypit on REDUX_OUT/LRIS_blue/long_600_4000_d560/lris_blue_long_600_4000_d560.pypit ---
~~~
So there are no negative entries in `edgearr` at this point in the `trace_slits` function, but I haven't figured out what this means in context yet.  I'm getting a similar failure when I try to run my DEIMOS tests.

**Comment by rcooke-ast** [comment]
Thanks for running the tests!

I was in the process of making some tweaks to the slittrace branch when my laptop died... (this is also why I haven't yet merged slittrace to the master branch). I'll be sure to update the code once I have my new laptop restored.

**Comment by profxj** [comment]
Where are we at with this PR?  Is it safe to merge?

**Comment by rcooke-ast** [comment]
I'm on my new computer and I don't have a copy of the tests (or good enough internet access to download it right now). I'll fix this when I'm back in the UK (in a week). Is this bug holding up progress?

**Comment by adwasser** [comment]
That's fine, I can work out some of the slitmask integration in the meantime.

**Comment by rcooke-ast** [comment]
OK, given that the slittrace algorithm works on the test suite, and on the trace frames that everyone sent me, I'm going to merge.

**Pull request merged**